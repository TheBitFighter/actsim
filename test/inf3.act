template <pint W, V, D>
defproc src (chan!(int<W>) SRC)
{
  int<W> tmp;
  int<1> dir;

  int cnt;

  chp {
    tmp := int(V, W);
    dir := 0;
    cnt := 0;
    *[ 
      [ cnt = 0 -> 
        SRC!tmp; 
        log("%xSOURCED VALUE: ", tmp); 
        tmp := dir = 0 ? tmp >> 1 : tmp << 1; 
        dir := ~dir;
        cnt := D != 0 ? cnt + 1 : 0
      []cnt = D & D != 0 -> 
        cnt := 0
      []else -> 
        cnt := cnt + 1
      ]
    ]
  }
}

template <pint W,D>
defproc buffer (chan(int<W>) IN; chan(int<W>) OUT)
{

  int<W> c[D];
  chan(int<W>) C[D];
  int<W> x;

  chp {
     *[ IN!0 ],
    *[ 
      [|#IN -> IN?c[0]; C[0]!c[0] |] 
    ],
    (,k:1..D-2: *[ C[k-1]?c[k]; C[k]!c[k] ] ),
    *[ C[D-2]?c[D-1]; OUT!c[D-1]],
    *[ OUT?x ]
  }
}

template <pint W>
defproc sink (chan?(int<W>) SINK)
{
  int<W> tmp;

  chp {
    tmp := 0;
    *[ SINK?tmp; log("%xSINKED VALUE: ", tmp) ]
  }
}

defproc test ()
{

//  src<8,8,0> src;
//  sink<8> sink;
  buffer<8,16> buffer;

//  src.SRC = buffer.IN;
//  buffer.OUT = sink.SINK;
}

test t;
