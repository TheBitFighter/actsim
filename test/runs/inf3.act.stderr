In expanding buffer<8,16> (inf3.act:29)
In expanding test (inf3.act:59)
In expanding ::<Global>
Error on or near line number 37.
WARNING: potential write conflict in CHP description
   Variable: `C[1]' (count=2)
   Variable: `C[2]' (count=2)
   Variable: `C[3]' (count=2)
   Variable: `C[4]' (count=2)
   Variable: `C[5]' (count=2)
   Variable: `C[6]' (count=2)
   Variable: `C[7]' (count=2)
   Variable: `C[8]' (count=2)
   Variable: `C[9]' (count=2)
   Variable: `C[10]' (count=2)
   Variable: `C[11]' (count=2)
   Variable: `C[12]' (count=2)
   Variable: `C[13]' (count=2)
WARNING: buffer<8,16>: substituting chp model (requested prs, not found)
